consul/data_source_consul_agent_self.go

```diff
@@ -170,6 +170,11 @@ func dataSourceConsulAgentSelf() *schema.Resource {
 	return &schema.Resource{
 		Read: dataSourceConsulAgentSelfRead,
 		Schema: map[string]*schema.Schema{
+			"deprecated": {
+				Type:       schema.TypeBool,
+				Required:   true,
+				Deprecated: "The consul_agent_self resource will be deprecated and removed in a future version. More information: https://github.com/terraform-providers/terraform-provider-consul/issues/46",
+			},
```

Comment:
We have precedent for deprecating resource types from back when some of them were reworked as data sources, implemented via [a transform function](https://github.com/hashicorp/terraform/blob/b8a1a77f6b2ec6b7d9502e2731e11a2b35657d76/helper/schema/data_source_resource_shim.go#L21-L41).

It could be reasonable to add another similar function that allows attaching a more free-form deprecation message via the same hook. The message added here, if non-empty, is returned as a warning during the validation step.

Your approach here is an interesting workaround, but it'll cause users to need to upgrade their configs to add this additional attribute, making this more of an "opt in to legacy behavior" flag. Certainly better than just removing it outright, of course!